developer_guides/fuzzing/testbench_afl_fuzzing.rst

```diff
@@ -0,0 +1,88 @@
+.. _testbench-afl-fuzzing:
+
+Build a Fuzzing Testbench with AFL
+##################################
+
+American fuzzy lop (AFL) is a free software fuzzer that can be used to detect software bugs. Use these instructions to build and run a testbench with AFL.
+
+Install AFL
+***********
+
+Follow the steps in the `AFL Quick Start Guide <https://github.com/google/AFL/blob/master/docs/QuickStartGuide.txt>`_ to install AFL on your system.
+
+We assume that AFL is installed at:
+
+::
+
+   $HOME/work/
+
+
+Build a testbench with AFL instrumentation
+******************************************
+
+According to AFL's `README <https://github.com/google/AFL/blob/master/README.md>`_, AFL is a "brute-force fuzzer coupled with an exceedingly
+simple but rock-solid instrumentation-guided genetic algorithm." Add
+instrumentation to the code before running a fuzzer on it in order to get
+good results.
+
+When you build AFL from the previous step, an ``afl-gcc`` executable is
+generated; this works as a companion tool that acts as a drop-in
+replacement for ``gcc`` or ``clang``. Before you build the testbench, make
+sure you are compiling code with ``afl-gcc`` in order to add instrumentation
+to the code. The ``host-build-all.sh`` script from the ``scripts/`` directory
+**does exactly this when you run it with the -f option.**
+
+.. Note::
+   By default, the ``host-build-all.sh`` script assumes you have installed AFL in the ``$HOME/work/ directory``. If you install AFL in any other directory, you must change the path in this script.
+
+Run AFL
+*******
+
+From the AFL directory, run AFL by entering the following:
+
+::
+
+   ./afl-fuzz -i testcase_dir -o findings_dir /path/to/program [...params...] @@
+
+AFL assumes that the inputs for the program you wish to fuzz are
+in the form of files. So, you must create a directory that contains these
+input files. This is the ``testcase_dir`` in the above command.
+
+Since you are fuzzing the testbench, the ``program`` here is testbench.
+
+``params`` are the different parameters of the program apart from the input
+file.
+
+``@@``: Each file from ``testcase_dir`` is substituted in place of this.
+As AFL continues to run, newly-generated testcases are placed in
+``testcase_dir``, and AFL is run again with those testcases.
+
+Example
+*******
+
+Use AFL to fuzz the volume component of the testbench. Use topology files as inputs and place the topology files of volume components in an ``inputs`` directory:
```

Comment:
Again a small nitpick.

I think it's better if we write 'Using AFL to fuzz the volume component of the testbench' on a separate line at the beginning. As this helps the reader in understanding what this example is trying to do.